scripts/zones/Aydeewa_Subterrane/mobs/Pandemonium_Warden.lua

```diff
@@ -1,194 +1,522 @@
-            else
-                handlePet(mob, newPet, oldPet, target, 1839)
+            -- ensure pets are there
+            pwFuncs.popPets(mob)
+            -- 'All avatars are summoned at once, and with them plus the lamps up, its hard to move your character.'
+            -- 'You will probably get locked in place and die from game mechanics alone.'
+            for i = 1, 8 do
+                -- during final phase, pets are always index 1 and astral flows are always index 0
+                local newPet = GetMobByID(petIDs[0][i])
+                local oldPet = GetMobByID(petIDs[1][i])
+                pwFuncs.handlePet(mob, newPet, oldPet, target, 790 + i)
+                newPet:setUnkillable(true)
+                newPet:setAutoAttackEnabled(false)
+                newPet:setMagicCastingEnabled(false)
+                newPet:setMobAbilityEnabled(false)
+            end
+        else
+            -- use non-dverger 2-hour at 50% hp of current phase
+            if phaseSpecialID > 0 then
+                local halfHP = mobPhaseHP[phase] / 2
+                if mobHP < halfHP then
+                    mob:useMobAbility(phaseSpecialID)
+                    mob:setLocalVar('usedSpecial', 1)
+                end
             end
-        end
-
-        -- Increment astral
-        mob:setLocalVar('astralFlow', astral + 1)
 
-    -- Or, at least make sure pets weren't drug off
-    else
-        --[[ Unused
-        for i = 1, 8 do
-            local pet = nil
-            if phase == 21 then
-                pet = pets[astral % 2][i]
-            else
-                pet = pets[phase % 2][i]
+            -- make sure pets weren't dragged off...
+            for i = 0, 1 do
+                for j = 1, 8 do
+                    local pet = GetMobByID(petIDs[i][j])
+                    if pet:isAlive() and not pet:isEngaged() then
+                        pet:updateEnmity(target)
+                    end
+                end
             end
         end
-        ]]--
     end
 
     -- Check for time limit, too
     if os.time() > depopTime and mob:actionQueueEmpty() then
-        for i = 0, 1 do
-            for j = 1, 8 do
-                if pets[i][j]:isSpawned() then
-                    DespawnMob(petIDs[i][j])
-                end
-            end
-        end
-
         DespawnMob(ID.mob.PANDEMONIUM_WARDEN)
+    --  printf('Timer expired at %i. Despawning Pandemonium Warden.', depopTime)
     end
 end
 
-entity.onMobDeath = function(mob, player, optParams)
-    player:addTitle(xi.title.PANDEMONIUM_QUELLER)
-
-    -- Despawn pets
-    for i = 0, 1 do
-        for j = 1, 8 do
-            if GetMobByID(petIDs[i][j]):isSpawned() then
-                DespawnMob(petIDs[i][j])
-            end
-        end
+entity.onMobDeath = function(mob, player, isKiller)
+    if player ~= nil then
+        player:addTitle(xi.title.PANDEMONIUM_QUELLER)
+        mob:showText(mob, ID.text.PW_END_OF_NOTHING)
     end
+
+    pwFuncs.despawnPets()
 end
```

Comment:
onMobDeath was changed to pass a table of options instead of isKiller. You want the death text behind that so it doesn't get spammed for each player in the party, noKiller is if it dies while unclaimed like from a DoT or something.
```lua
entity.onMobDeath = function(mob, player, optParams)
    if optParams.isKiller or optParams.noKiller then
        mob:showText(mob, ID.text.PW_END_OF_NOTHING)
    end

    if player ~= nil then
        player:addTitle(xi.title.PANDEMONIUM_QUELLER)
    end

    pwFuncs.despawnPets()
end
```

Reply:
Nice. I remember reading about those changes but this is my first time using the new properties.

Also on wings i wasn't using `mob:showText` so i was able to run it for every player. Good catch there.